### PR TITLE
Fix undefined behavior in Rc/Arc allocation

### PR DESCRIPTION
**Pull request by murarth**
Manually calculate allocation layout for `Rc`/`Arc` to avoid undefined behavior

Closes #54908

### CODE REVIEW

**Comment by RalfJung** [comment]
Could you add a comment saying why we do it this way and not the old way?

Other than that, I agree the alignment problem is gone. However, I do not know nearly enough about layouts to say if that's a correct way to compute these things. Assigning to someone who knows more (feel free to pick someone else).

r? @eddyb

**Comment by gnzlbg** [comment]
cc @SimonSapin @Amanieu

**Comment by TimNN** [comment]
Ping from triage! This PR requires your review @eddyb. Also this PR was references from https://github.com/rust-rfcs/unsafe-code-guidelines/issues/35, is it blocked on that issue?

**Comment by eddyb** [comment]
This is not a compiler layout, but I suspect it's correct.

r? @alexcrichton

File: src/liballoc/rc.rs [hunk 1]
```diff
@@ -665,15 +665,14 @@ impl Rc<dyn Any> {
 impl<T: ?Sized> Rc<T> {
     // Allocates an `RcBox<T>` with sufficient space for an unsized value
     unsafe fn allocate_for_ptr(ptr: *const T) -> *mut RcBox<T> {
-        // Create a fake RcBox to find allocation size and alignment
-        let fake_ptr = ptr as *mut RcBox<T>;
-
-        let layout = Layout::for_value(&*fake_ptr);
+        // Calculate layout using the given value
+        let (layout, _) = Layout::new::<RcBox<()>>()
+            .extend(Layout::for_value(&*ptr)).unwrap();
```
Comment (RalfJung) [inline src/liballoc/rc.rs]: Please add a comment saying why casting the ptr and calling `Layout::for_value` on it does not work -- let's make sure we do not reintroduce this problem later again.

**Comment by alexcrichton** [comment]
Thanks for the PR and sorry for the delay! This looks good to go to me with @RalfJung's comment

**Comment by alexcrichton** [comment]
If possible, it'd also be great to have a regression test for this!

**Comment by RalfJung** [comment]
> If possible, it'd also be great to have a regression test for this!

We'll have on in miri. I have no idea how to make one with rustc only.

**Comment by murarth** [comment]
I've added a more detailed comment and the tests have passed.

**Comment by RalfJung** [comment]
Hm, this RFC made miri fail even if validation is turned off. Still investigating.

**Comment by RalfJung** [comment]
The error is that an allocation has size 32 but gets deallocated with size 28.

I assume the layout computations done here are incorrect.

**Pull request merged**